libplatsupport/src/mach/imx/gpt.c

```diff
@@ -217,9 +217,22 @@ static void gpt_handle_irq(void *data, ps_irq_acknowledge_fn_t acknowledge_fn, v
 
 uint64_t gpt_get_time(gpt_t *gpt)
 {
+    // Rollover of 32-bit counter can happen while we are reading it,
+    // We need to read in the following order (volatile modifier on gpt_map should ensure this):
+    // rollover_status bit
+    // gptcnt
+    // rollover_status bit.
+    // If rollover_status bit has changed, then the rollover happened sometime between the first
+    // read and the third read.  In this case conclude that the low_bits are 0.
+    uint32_t rollover_st = gpt->gpt_map->gptsr & BIT(ROV);
     uint32_t low_bits = gpt->gpt_map->gptcnt;
+    uint32_t rollover_st2 = gpt->gpt_map->gptsr & BIT(ROV);
```

Comment:
I wonder if this is enough to catch all corner cases, don't we need something like this:
```
hi_prev = gpt->high_bits
rollover_st = gpt->gpt_map->gptsr & BIT(ROV)
low_bits = gpt->gpt_map->gptcnt
gpt->gpt_map->gptsr & BIT(ROV);
high_bits = gpt->high_bits
if (hi_prev != high_bits) { 
    // there was a high_bits update in between
    low_bits  = 0; 
else if (rollover_st != rollover_st2) && (1 == rollover_st2) { 
    // low bits roll over detected before ISR could update gpt->high_bits
    high_bits++; 
}
```

Then

Reply:
The driver's api functions aren't thread safe and the driver requires that what ever is calling into it isn't calling functions concurrently.  Because of this, the value of gpt->high_bits isn't allowed to change within the scope of `gpt_get_time` as it's regular memory and only incremented by the driver.

Reply:
I think that should be in the comment also to explain why this is really sufficient here.

Reply:
I updated the comments.